### PR TITLE
feat(gpu): #1831 M3 — Q8_0 kernels + qwen35 loader self-check (task 2 infra)

### PR DESCRIPTION
**Pull request by bong-water-water-bong**
### **User description**
**Part of #1831** (open). M3 task-2 infrastructure: the raw-Q8_0 device path + kernels the qwen35moe decode needs, plus a hardware self-check.

**Kernels** (`src/hip_1bp_kernels.hip`):
- `h1bp_q8gemv_kernel` — Q8_0 (GGUF dtype 8: fp16 scale + 32 int8 per block, 34 B/block) row gemv, mirroring `h1bp_gemv_kernel`.
- `h1bp_q8embed_kernel` — Q8_0 embedding row dequant.
- `h1bp_q8gemv_slice_kernel` — fused-expert per-expert gemv over expert-major contiguous tensors (the MUL_MAT_ID equivalent confirmed in the M3 capture).

**Model format note:** the staged `Qwen3.6-35B-A3B-Q8_0.gguf` is **Q8_0** (llama ftype + GGUF dtype 8), not Q8_K — block-32 geometry per the #2104 dtype table. Dequant-to-f32 of the whole 35B is infeasible on-device, so decode will run on raw Q8_0 tiles.

**Self-check** (`H1BP_Q35_SELFCHECK=1`, runs only after the M2 structural gate, dense path untouched): pulls `blk.0.attn_qkv` raw (34 B/block), runs the device Q8 gemv (8192×2048) and compares row 0 vs a host dequant+dot — verifies loader + kernel on real hardware before any decode wiring. qwen35 decode still refused (tasks 3-5).

Run on strixhalo: `H1BP_Q35_SELFCHECK=1 ./build/bench_hip_1bp ~/models/Qwen3.6-35B-A3B-Q8_0.gguf` → expect `selfcheck ... row0 ref=… dev=… → PASS` then the structure-verified message.

___

### **PR Type**
Enhancement

___

### **Description**
- Adds full qwen35moe eager decode to HIP backend.
  - GatedDeltaNet linear attention, fused QKV, conv1d.
  - Full-attn layers with RoPE and KV cache.
  - MoE top-8 router, gated/shared experts.

- Introduces Q8_0 gemv/embed/slice kernels and qwen35 decode kernels.

- Loader/self-check behind H1BP_Q35_LOAD and H1BP_Q35_SELFCHECK.

- Bench tool feeds prompts with H1BP_Q35_PROMPT.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  GGUF["Qwen3.6-35B-A3B Q8_0 GGUF"] --> Loader["H1BP_Q35_LOAD device loader"]
  Loader --> Decode["qwen35_step eager decode"]
  Decode --> GDN["GDN: conv1d + delta rule"]
  Decode --> FA["Full-attn: RoPE + KV cache"]
  Decode --> MoE["MoE top-8 + shared expert"]
  GDN --> Head["output norm + lm_head argmax"]
  FA --> Head
  MoE --> Head
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>backend_hip_1bp.cpp</strong><dd><code>qwen35moe loader, scratch, and eager decode</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/backend_hip_1bp.cpp

<ul><li>Adds qwen35moe device-weight structs (<code>Q35L</code>) and scratch buffers for <br>conv, recurrent delta, KV cache, and MoE state.<br> <li> Adds env-gated raw-Q8_0 loader (<code>H1BP_Q35_LOAD</code>) and device gemv <br>self-check (<code>H1BP_Q35_SELFCHECK</code>).<br> <li> Implements <code>qwen35_step()</code> eager decode covering GDN, full-attention, <br>and MoE paths; dispatches from <code>generate_fast()</code>.<br> <li> Extends <code>reset()</code>/<code>destroy()</code> to manage q35 state and device buffers.</ul>

</details>

  </td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2127/files#diff-c4812ee96d278d2d2b452760b359c0521f1a016dce68e46c6668555b33eec182">+412/-1</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>hip_1bp_kernels.hip</strong><dd><code>Q8_0 gemv and qwen35 decode kernels</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/hip_1bp_kernels.hip

<ul><li>Adds Q8_0 row gemv, embedding dequant, and expert-slice gemv kernels.<br> <li> Implements GDN kernels: causal conv1d+silu, head L2-norm, and <br>delta-rule recurrence.<br> <li> Adds full-attention kernel over the per-layer KV cache plus RoPE.<br> <li> Adds MoE router/top-k, sigmoid/silu gating, scale, accumulate, and <br>elementwise kernels.</ul>

</details>

  </td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2127/files#diff-e7c676fbc5592cbbf17f1c7db3df89e539cc6bfef40bfc8014de1d807ac66396">+258/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>bench_hip_1bp.cpp</strong><dd><code>bench prompt-mode support for qwen35 validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tools/bench_hip_1bp.cpp

<ul><li>Adds <code>H1BP_Q35_PROMPT</code> support to feed prompt tokens through <code>generate()</code> <br>before decode benchmarking.<br> <li> Skips reset/token restart and returns early in prompt mode for <br>accurate decode measurement.</ul>

</details>

  </td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2127/files#diff-ba7085f2c60678386a91382bf78ea72395c9d7dc1345f2dbb5dc698687d91522">+15/-4</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by bong-water-water-bong** [comment]
**M3 status — decode engine validated vs in-repo golden.** The HIP qwen35moe decode path (full-attn + GDN + MoE kernels behind H1BP_Q35_TRY) now matches the host-side golden formula at **corr 1.00000 across all 40 layers** (single token, per-layer hidden states) and **corr 0.99884 / argmax-identical** on a stateful 2-token run (conv-state + delta-recurrence + full-attn KV across tokens). Q8_0 (34 B/block) dequant confirmed arithmetically; final-logits magnitude matches llama-HIP (2.3 vs 2.26). Decode stays env-gated (H1BP_Q35_LOAD + H1BP_Q35_TRY), so default behavior is unchanged (dense models + qwen35 refusal intact).
